Add equal height row component (#5039)
canonical/vanilla-framework: templates/docs/patterns/equal-height-row.md

In addition to enforcing column grid items alignment within each row, responsive behaviour is provided out of the box as shown below:

<table>
  <thead>
    <tr>
      <th style="width: 50ch">Screen size (px)</th>
      <th>Behaviour</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Less than <code>$breakpoint-small</code></td>
      <td>Each column spans across the entire parent grid. Column items are vertically stacked.</td>
    </tr>
    <tr>
      <td><code>$breakpoint-small</code> - <code>$breakpoint-large</code></td>
      <td>Each column spans across the entire parent grid. The first item within each column is placed on the left of the other column items.</td>
    </tr>
    <tr>
      <td>Greater than <code>$breakpoint-large</code></td>
      <td>Columns within the row are displayed horizontally. Column items are vertically stacked.</td>
    </tr>
  </tbody>
</table>

This component is an extension of the [grid component](/docs/patterns/grid). There are several usage variations detailed in the examples below.

> ## Four items per column and full-screen dividers

Reviewer: To make it feel less like an exception let's simplify this title to "Full example" maybe

```suggestion
## Full example
```

> Each column can have up to 4 items.

Reviewer: ```suggestion
Each column of the component can have up to 4 sub-grid items that will keep equal heights between the columns.
```

> You can also insert dividers that span across all columns within a row using `.has-1st-divider`, `.has-2nd-divider` and `.has-3rd-divider`.

Reviewer: I'd suggest moving the dividers documentation below the "Full example".

Let's make a separate `### Cross-column dividers` heading and keep the dividers documentation and notifications under it, below the embedded example.

Author: Yeah that definitely makes more sense, thanks!

---
wrapper_template: '_layouts/docs.html'
context:
  title: Equal height row | Components
---

The equal height row component aims to provide consistent alignment for grid items within a row format. This is achieved using the CSS `subgrid` feature which allows column grids to share the same layout used in the parent row grid.

This component is an extension of the [grid component](/docs/patterns/grid). There are two ways to use this component:

> 1. Used as a four column row grid (`.p-equal-height-row`). Each column (`.p-equal-height-row__col`) within the row is a css grid, and may have up to four grid items (`.p-equal-height-row__item`).

Reviewer: ```suggestion
1. Used as a four column row grid (`.p-equal-height-row`) spanning the whole width of the default page grid. Each column (`.p-equal-height-row__col`) within the component is a sub-grid may have up to four items (`.p-equal-height-row__item`).
```

Reviewer: Actually, I think this description can be moved inside the "Four column row" section, above the example.

> 2. Nest inside the `25-75` grid pattern on large screen size (`.row--25-75-on-large`). The row (`.p-equal-height-row`) should be placed within the 75% container with a maximum of three columns.

Reviewer: ```suggestion
2. Nested inside the 25/75 grid split pattern on large screen size (`.row--25-75-on-large`). The row (`.p-equal-height-row`) should be placed within the 75% container with a maximum of three columns.
```

This also could be moved down inside the "Three column row" as a description above the example.

In addition to enforcing column grid items alignment within each row, responsive behaviour is provided out of the box as shown below:

<table>
  <thead>
    <tr>
      <th style="width: 50ch">Screen size (px)</th>
      <th>Behaviour</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Less than <code>$breakpoint-small</code></td>
      <td>Each column spans across the entire parent grid. Column items are vertically stacked.</td>
    </tr>
    <tr>
      <td><code>$breakpoint-small</code> - <code>$breakpoint-large</code></td>
      <td>Each column spans across the entire parent grid. The first item within each column is placed on the left of the other column items.</td>
    </tr>
    <tr>
      <td>Greater than <code>$breakpoint-large</code></td>
      <td>Columns within the row are displayed horizontally. Column items are vertically stacked.</td>
    </tr>
  </tbody>
</table>

The below sections provides examples for using this component to achieve common design patterns.

> ## Four column row pattern

Reviewer: I think the "pattern" word in the titles is not necessary

```suggestion
## Four column row
```

</a></div>

## Full screen dividers

You can insert dividers that span across all columns within a row using `.has-1st-divider`, `.has-2nd-divider` and `.has-3rd-divider`.

<div class="p-notification--information">
  <div class="p-notification__content">
    <h5 class="p-notification__title">Note:</h5>
    <p class="p-notification__message">For smaller screen sizes i.e. <code>< $breakpoint-large</code> the divider will appear below relevant items within each column.</p>
  </div>
</div>

<div class="p-notification--caution">
  <div class="p-notification__content">
    <p class="p-notification__message">You may only have two dividers maximum. For example, if you have <code>.has-1st-divider</code> and <code>.has-2nd-divider</code> set for the row, then <code>.has-3rd-divider</code> will not be shown. This is a limitation due to usage of pseudo elements for visualising dividers that is capable of spanning across grid gaps.</p>
  </div>
</div>

<div class="embedded-example"><a href="/docs/examples/patterns/equal-height-row/full-screen-dividers/" class="js-example">
View example of full screen dividers in equal height row component.
</a></div>

> ## 25-75 grid pattern

Reviewer: Let's call is "Three column row" and move it up below "Four column row"

Reviewer: ```suggestion
## Three column row
```
